src/passes/OptimizeInstructions.cpp

```diff
@@ -907,6 +907,29 @@ struct OptimizeInstructions
       }
     }
 
+    if (curr->op == ExtendUInt32 || curr->op == ExtendSInt32) {
+      if (auto* load = curr->value->dynCast<Load>()) {
+        // i64.extend_i32_s(i32.load(_8|_16)(_u|_s)(x))  =>
+        //    i64.load(_8|_16|_32)(_u|_s)(x)
+        //
+        // i64.extend_i32_u(i32.load(_8|_16)(_u|_s)(x))  =>
+        //    i64.load(_8|_16|_32)(_u|_s)(x)
+        if (curr->op == ExtendUInt32 && load->signed_ && load->bytes <= 2) {
+          // Skip special two special cases:
+          //   i64.extend_i32_u(i32.load8_s(x)) and
+          //   i64.extend_i32_u(i32.load16_s(x))
+        } else if (!load->isAtomic) {
```

Comment:
The isAtomic can be moved to the outer if, I think, which would simplify this. If it's atomic we don't need to look into this new code at all.

But: why can't we handle atomic loads here?

Reply:
> why can't we handle atomic loads here?

I's also not sure about this. But I know for similar reinterpret + load rule we can't use atomics and this was found by fuzzer. Also I checked what LLVM do it similar case and it avoid optimization ass well.

Reply:
Hmm, however Rust optimize this with atomics:
https://godbolt.org/z/b94GYTYPK

So probably we could do this as well.